### PR TITLE
Update incorrect data processing in DataCollatorForChatML

### LINKED ISSUE #2169: Incorrect data processing in DataCollatorForChatML
### System Info

Python 3.11.9
trl 0.11.0
transformers 4.45.1

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

```python
from trl.trainer.utils import DataCollatorForChatML
from transformers import AutoTokenizer
# base model settings
tokenizer = AutoTokenizer.from_pretrained("codellama/CodeLlama-7b-Instruct-hf")
tokenizer.pad_token = tokenizer.bos_token if tokenizer.pad_token is None else tokenizer.pad_token

# Initialize the data collator for processing the chat model inputs
max_length = 1024
ignore_index = -100
messages_key = "messages"
examples = [
    {
        messages_key:
            [
                {"role": "user", "content": "Does the following code contain any security vulnerabilities? Return true or false.\nchar buffer[10];\nchar input[50];\nstrcpy(buffer, input);\n"}, 
                {"role": "assistant", "content": "true"}
            ]
    }
]
collator = DataCollatorForChatML(tokenizer=tokenizer, max_length=max_length, ignore_index=ignore_index, messages_key=messages_key)

# Process the data
data = collator(examples)
print(data["input_ids"])
print(data["labels"])
```

# Founded Bugs and Solutions

## Output from Demo Script

```python
>>> print(data["input_ids"])
tensor([[    1,     1,   518, 25580, 29962,  5538,   278,  1494,   775,  1712,
           738,  6993, 23180, 11614, 29973,  7106,  1565,   470,  2089, 29889,
            13,  3090,  6835, 29961, 29896, 29900,  1385,    13,  3090,  1881,
         29961, 29945, 29900,  1385,    13,   710, 23141, 29898,  9040, 29892,
          1881,   416,   518, 29914, 25580, 29962, 29871,     1,  1565]])
>>> print(data["labels"])
tensor([[-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,    1,
            2]])
```

## Founded Bugs

Preliminary: the expected output from assistant is a string "true" and its token id is 1565. The bos token id is 1 and the eos token id is 2.

1. Extra bos token and no eos token in data["input_ids"]. We can see that data["input_ids"] got an extra bos token at the beginning (i.e., [1, 1, ...]), and at the end of data["input_ids"] we can see another extra bos token (i.e., [..., 1, 1565]) and no eos token.
2. No target string in data["labels"]. As for the data["labels"], it does not include the expected output from assistant ("true" token) and only the bos and eos token is preserved (i.e., [..., -100, 1, 2]).

## Expected Output of Demo Script

The expected output should look like this:

```python
>>> print(data["input_ids"])
tensor([[    1,   518, 25580, 29962,  5538,   278,  1494,   775,  1712,   738,
          6993, 23180, 11614, 29973,  7106,  1565,   470,  2089, 29889,    13,
          3090,  6835, 29961, 29896, 29900,  1385,    13,  3090,  1881, 29961,
         29945, 29900,  1385,    13,   710, 23141, 29898,  9040, 29892,  1881,
           416,   518, 29914, 25580, 29962, 29871,  1565, 29871,     2]])
>>> print(data["labels"])
tensor([[ -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  1565, 29871,     2]])
```

We can see the expected output of data["input_ids"] and data["labels"] should not have any extra bos tokens, and the data["labels"] preserved the target string correctly (i.e., [..., -100, 1565, 29871, 2]).
I believe this result is more compatible with the loss calculation like LlamaForCausalLM as demonstrated below:

```python
loss = None
if labels is not None:
    # Upcast to float if we need to compute the loss to avoid potential precision issues
    logits = logits.float()
    # Shift so that tokens < n predict n
    shift_logits = logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()
    # Flatten the tokens
    loss_fct = CrossEntropyLoss()
    shift_logits = shift_logits.view(-1, self.config.vocab_size)
    shift_labels = shift_labels.view(-1)
    # Enable model parallelism
    shift_labels = shift_labels.to(shift_logits.device)
    loss = loss_fct(shift_logits, shift_labels)
```

## Location and Solution to the Bugs

The bugs is located at the beginning of __call__ function of DataCollatorForChatML:

```python
prompts = []
completions = []

for example in examples:
    messages = example[self.messages_key]
    formatted_chat = self.tokenizer.apply_chat_template(messages, tokenize=False)

    # Split the formatted chat into prompt and completion
    assistant_messages = [msg for msg in messages if msg["role"] == "assistant"]
    last_assistant_message = assistant_messages[-1]["content"]
    prompt = formatted_chat.rsplit(last_assistant_message, 1)[0]
    completion = last_assistant_message

    prompts.append(prompt)
    completions.append(completion)

# Tokenize prompts and completions
tokenized_prompts = self.tokenizer(
    prompts, truncation=True, max_length=self.max_length, padding=False, return_tensors=None
)
tokenized_completions = self.tokenizer(
    completions, truncation=True, max_length=self.max_length, padding=False, return_tensors=None
)

# Combine prompts and completions
input_ids = []
attention_mask = []
labels = []

for prompt, completion in zip(tokenized_prompts["input_ids"], tokenized_completions["input_ids"]):
    combined_input_ids = prompt + completion
    combined_attention_mask = [1] * len(combined_input_ids)

    # Create labels for one-token ahead task, masking the prompt
    combined_labels = [self.ignore_index] * len(prompt) + completion[:-1]
    combined_labels.append(self.tokenizer.eos_token_id)  # Add EOS token as final target

    input_ids.append(combined_input_ids)
    attention_mask.append(combined_attention_mask)
    labels.append(combined_labels)
```

### Reason for Bug1

The code that raises bug1 is due to:

```python
# Tokenize prompts and completions
tokenized_prompts = self.tokenizer(
    prompts, truncation=True, max_length=self.max_length, padding=False, return_tensors=None
)
tokenized_completions = self.tokenizer(
    completions, truncation=True, max_length=self.max_length, padding=False, return_tensors=None
)
```

Because prompts are obtained from formatted_chat, and formatted_chat already adds bos token by default:

```python
formatted_chat = self.tokenizer.apply_chat_template(messages, tokenize=False)
```

So, after using self.tokenizer to tokenize prompts and completions, it will add extra bos token to them.

### Fix to Bug1

Set the default parameter of add_special_token from True to False in the following code:

```python
# Tokenize prompts and completions
tokenized_prompts = self.tokenizer(
    prompts, truncation=True, max_length=self.max_length, padding=False, return_tensors=None, add_special_token=False
)
tokenized_completions = self.tokenizer(
    completions, truncation=True, max_length=self.max_length, padding=False, return_tensors=None, add_special_token=False
)
```

### Reason for Bug2

The location that raises bug2 is here:

```python
completion = last_assistant_message
```

Since apply_chat_template will add eos token to formatted_chat, above code does not include the eos token to the completion correctly. And 

```python
tokenized_completions = self.tokenizer(
    completions, truncation=True, max_length=self.max_length, padding=False, return_tensors=None
)
```

will adds extra bos token to the tokenized_completions, so in this code:

```python
combined_labels = [self.ignore_index] * len(prompt) + completion[:-1]
```

the expected output from assistant is ignored and only the bos token is retained.

### Fix to Bug2

The correct implementation should be:

```python
completion = last_assistant_message + formatted_chat.rsplit(last_assistant_message, 1)[1].
```

### PR DESCRIPTION
**Pull request by ruijunfeng**
# What does this PR do?
Fix the extra BOS token and the absence of an EOS token in the returned input_ids, and potentially the absence of a target string in the returned labels.

Fixes #2169

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case. (https://github.com/huggingface/trl/issues/2169)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by kashif** [comment]
awesome @ruijunfeng can we also have a test for this?

**Comment by ruijunfeng** [comment]
> awesome @ruijunfeng can we also have a test for this?

Sure thing, I have tested it on Llama2 series and gemma1 series with my own dataset, and it seems to work well. Let me know if you need me to provide anything.

**Comment by kashif** [comment]
sorry for the misunderstanding, I meant something like:

```python
class TestDataCollatorForChatML(unittest.TestCase):
    def setUp(self):
        self.tokenizer = AutoTokenizer.from_pretrained("gpt2")
        self.tokenizer.pad_token = self.tokenizer.eos_token
        self.collator = DataCollatorForChatML(tokenizer=self.tokenizer, max_length=20)

    def test_data_collator(self):
        examples = [
            {
                "messages": [
                    {"role": "user", "content": "Hello!"},
                    {"role": "assistant", "content": "Hi there! How can I help you today?"},
                    {"role": "user", "content": "What's the weather like?"},
                    {"role": "assistant", "content": "I'm sorry, but I don't have access to real-time weather information."},
                ]
            },
            {
                "messages": [
                    {"role": "user", "content": "Tell me a joke."},
                    {"role": "assistant", "content": "Why don't scientists trust atoms? Because they make up everything!"},
                ]
            }
        ]

        batch = self.collator(examples)

        self.assertIn("input_ids", batch)
        self.assertIn("attention_mask", batch)
        self.assertIn("labels", batch)
        self.assertIn("prompts", batch)
        self.assertIn("prompt_attention_mask", batch)

        self.assertEqual(batch["input_ids"].shape[0], 2)
        self.assertEqual(batch["attention_mask"].shape[0], 2)
        self.assertEqual(batch["labels"].shape[0], 2)
        self.assertEqual(batch["prompts"].shape[0], 2)
        self.assertEqual(batch["prompt_attention_mask"].shape[0], 2)

        # Check if the shapes are consistent
        self.assertEqual(batch["input_ids"].shape, batch["attention_mask"].shape)
        self.assertEqual(batch["input_ids"].shape, batch["labels"].shape)
        self.assertEqual(batch["prompts"].shape, batch["prompt_attention_mask"].shape)

        # Check if the prompts are shorter than or equal to the full input
        self.assertTrue((batch["prompts"].shape[1] <= batch["input_ids"].shape[1]).all())
```

so we can explicitly check for the incorrect data processing and the fix you so kindly provided

**Comment by ruijunfeng** [comment]
Hi there, I have run your test code, and I think your test code has a small mistake. You are using the tokenizer for GPT-2: 
```python
self.tokenizer = AutoTokenizer.from_pretrained("gpt2")
```
However, GPT-2 does not have a default chat_template, so it will cause an error in this line of DataCollartorForChatML
```python
self.tokenizer.apply_chat_template(messages, tokenize=False). 
```
I believe the correct way to test this is by manually setting the chat_template for the tokenizer, like this in your setup function:
```python
tokenizer.chat_template = "{{ bos_token }}{% for message in messages %}{{ message['role'] }}: {{ message['content'] }}{% endfor %}{{ eos_token }}". 
```
Alternatively, you could use a model that has been fine-tuned on instructions, such as Llama-Instructed, whose tokenizer has a default chat_template.

**Comment by kashif** [comment]
sorry again for the misunderstanding, what i wanted to say is that you can use the above as a template to write the tests in your PR

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -265,17 +265,17 @@ def __call__(self, examples: List[Dict[str, Any]]) -> Dict[str, torch.Tensor]:
...
             last_assistant_message = assistant_messages[-1]["content"]
             prompt = formatted_chat.rsplit(last_assistant_message, 1)[0]
-            completion = last_assistant_message
+            completion = last_assistant_message + formatted_chat.rsplit(last_assistant_message, 1)[1]
 
             prompts.append(prompt)
             completions.append(completion)
 
         # Tokenize prompts and completions
         tokenized_prompts = self.tokenizer(
```
Comment (lewtun) [inline trl/trainer/utils.py:274]: Let's add a comment here to remind ourselves why we set `add_special_tokens=False`
```suggestion
        # We assume the inputs are already formatted with ChatML, so extra BOS/EOS tokens should not be added.
        tokenized_prompts = self.tokenizer(
```

**Review by lewtun — COMMENTED** [review]
Thanks a lot for the fix @ruijunfeng ! Would you mind adding a unit test which validates the fix works as expected? This will also help ensure future regressions don't leak into the codebase :)

**Comment by kashif** [comment]
@lewtun i added a test that fails on main and passes here and @ruijunfeng I pushed it into your PR

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@ruijunfeng kindly run the `pre-commit` in the root dir of TRL to fix formatting

File: tests/test_utils.py [hunk 2]
```diff
@@ -169,3 +175,83 @@ def test_val_none(self):
...
+        labels = data["labels"][0].tolist()
+
+        # Expected tokens
+        expected_bos = self.bos_token_id
+        expected_eos = self.eos_token_id
+        expected_assistant_token = self.assistant_output_token_id
+
+        # Verify that input_ids start with a BOS token and there are no extra ones
+        self.assertEqual(input_ids[0], expected_bos, "The first token of input_ids should be BOS token.")
+        self.assertNotEqual(input_ids[1], expected_bos, "The second token of input_ids should not be BOS token (extra BOS).")
```
Comment (kashif) [inline tests/test_utils.py:235]:
```suggestion
        self.assertNotEqual(
            input_ids[1], expected_bos, "The second token of input_ids should not be BOS token (extra BOS)."
        )
```

File: tests/test_utils.py [hunk 3]
```diff
@@ -169,3 +175,83 @@ def test_val_none(self):
...
+        last_assistant_response = self.examples[0][self.messages_key][-1]["content"]
+        last_assistant_response_tokens = self.tokenizer.encode(last_assistant_response, add_special_tokens=False)
+
+        # Find the start and end of the last assistant's response in the labels
+        response_start = next(i for i, label in enumerate(labels) if label != self.ignore_index)
+        response_end = next(i for i in range(len(labels) - 1, -1, -1) if labels[i] != self.ignore_index)
+
+        actual_response = labels[response_start : response_end - 1]
+        self.assertEqual(
+            actual_response, last_assistant_response_tokens, "The labels should preserve the last assistant's response tokens."
```
Comment (kashif) [inline tests/test_utils.py:253]:
```suggestion
            actual_response,
            last_assistant_response_tokens,
            "The labels should preserve the last assistant's response tokens."
```

**Comment by ruijunfeng** [comment]
@kashif and @lewtun, thank you both for adding the tests and comments. I’ve double-checked the tests and made updates to the assert statements and comments to improve consistency and clarity. Additionally, I noticed that the current check for the EOS token in input_ids only verifies its presence. I have modified it to ensure that the last token of input_ids is the EOS token for a more thorough check.

**Comment by kashif** [comment]
yes i didnt take that into account

**Comment by kashif** [comment]
@qgallouedec fixed the test taking padding into account

File: tests/test_utils.py [hunk 4]
```diff
@@ -169,3 +176,80 @@ def test_val_none(self):
         assert "my_model" in card_text
         assert 'pipeline("text-generation", model="username/my_hub_model", device="cuda")' in card_text
         assert "My Trainer" in card_text
+
+
+class TestDataCollatorForChatML(unittest.TestCase):
+    def setUp(self):
+        # Initialize the tokenizer
+        self.tokenizer = AutoTokenizer.from_pretrained("codellama/CodeLlama-7b-Instruct-hf")
```
Comment (qgallouedec) [inline tests/test_utils.py:184]: For tests, I'd prefer a smaller model. Or is there something specific with this one?

File: tests/test_utils.py [hunk 5]
```diff
@@ -169,3 +176,80 @@ def test_val_none(self):
...
+        prompt_only = data["prompts"][0].tolist()
+
+        # Expected tokens
+        expected_bos = self.bos_token_id
+        expected_eos = self.eos_token_id
+
+        # Verify that input_ids start with optional padding tokens  and a single BOS token and there are no extra ones
+        first_non_pad = next(token for token in input_ids if token != self.tokenizer.pad_token_id)
+        self.assertEqual(first_non_pad, expected_bos, "The first non-padding token of input_ids should be BOS token.")
+        bos_indices = [i for i, token in enumerate(input_ids) if token == expected_bos]
```
Comment (qgallouedec) [inline tests/test_utils.py:223]:
```suggestion
        bos_indices = [i for i, token in enumerate(input_ids) if token == self.bos_token_id]
```

File: tests/test_utils.py [hunk 6]
```diff
@@ -169,3 +176,78 @@ def test_val_none(self):
...
+            max_length=self.max_length,
+            ignore_index=self.ignore_index,
+        )
+
+    def test_data_collator_for_chatml(self):
+        # Process the data
+        data = self.collator(self.examples)
+
+        # Decode input_ids and labels for verification
+        input_ids = data["input_ids"][0].tolist()
```
Comment (qgallouedec) [inline tests/test_utils.py:212]: For some reason, we get an extra token (`29871`) with the collator. Any idea why?

```python
>>> input_ids[15:]  # 15 first tokens are padding
[1, 518, 25580, 29962, 1724, 338, 2253, 1135, 22769, 29973, 518, 29914, 25580, 29962, 29871, 25685, 29889, 29871, 2]
>>> self.tokenizer(self.tokenizer.apply_chat_template(self.examples[0]["messages"], tokenize=False), add_special_tokens=False))
print(self.tokenizer(self.tokenizer.apply_chat_template(self.examples[0]["messages"], tokenize=False), add_special_tokens=False)
[1, 518, 25580, 29962, 1724, 338, 2253, 1135, 22769, 29973, 518, 29914, 25580, 29962, 25685, 29889, 29871, 2]
```

**Comment by kashif** [comment]
@qgallouedec I have:
```
input_ids[15:]  # 15 first tokens are padding
[1, 518, 25580, 29962, 1724, 338, 2253, 1135, 22769, 29973, 518, 29914, 25580, 29962, 25685, 29889, 29871, 2]
(Pdb) self.tokenizer(self.tokenizer.apply_chat_template(self.examples[0]["messages"], tokenize=False), add_special_tokens=False)
{'input_ids': [1, 518, 25580, 29962, 1724, 338, 2253, 1135, 22769, 29973, 518, 29914, 25580, 29962, 25685, 29889, 29871, 2], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
```

File: tests/test_utils.py (hunk 4 above)
Comment (qgallouedec) [reply tests/test_utils.py:184]: We just just load the tokenizer anyway. Not a big deal then.

**Comment by kashif** [comment]
@ruijunfeng can you kindly check with the current refactoring of the datacollator, I have simplified it

**Comment by ruijunfeng** [comment]
@kashif Hi there, I still found a small bug in the refactor code. I used the dataset in your unit test and print out the results like this:
```python
>>> tokenizer.decode(data["prompts"][0])
'<s><s><s><s><s><s><s>user: What is better than ugly?</s>'
>>> tokenizer.decode(data["input_ids"][0])
'<s><s><s><s><s><s><s><s><s><s><s><s><s><s><s><s>user: What is better than ugly?assistant: Beautiful.</s>'
>>> data["labels"][0, -6:]
tensor([ -100, 22137, 29901, 25685, 29889,     2])
>>> tokenizer.decode(data["labels"][0, -5:])
'istant: Beautiful.</s>'
```
Seems like the labels mistakenly wrap part of the "assistant: " and the prompts has missed the "assistant: ".

**Comment by kashif** [comment]
just trying to reproduce this on my end, I have as output from the data collator:

```
self.tokenizer.decode(input_ids[0])
'</s></s></s></s></s></s></s></s></s></s></s></s></s></s></s><s> [INST] What is better than ugly? [/INST] Beautiful. </s>'
```

```
self.tokenizer.decode(prompts_input_ids[0])
'</s></s></s></s></s></s><s> [INST] What is better than ugly? [/INST]'
```

and the labels are only set for the completion:
```
labels[0]
tensor([ -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
         -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
         -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100, 25685,
        29889, 29871,     2])
```

```
self.tokenizer.decode(labels[0][-4:])
'Beautiful. </s>'
```
at which point are you printing the data from?

File: tests/test_utils.py [hunk 7]
```diff
@@ -169,3 +176,78 @@ def test_val_none(self):
...
+        labels = data["labels"][0].tolist()
+        prompt_only = data["prompts"][0].tolist()
+
+        # Verify that input_ids start with optional padding tokens  and a single BOS token and there are no extra ones
+        first_non_pad = next(token for token in input_ids if token != self.tokenizer.pad_token_id)
+        self.assertEqual(
+            first_non_pad, self.bos_token_id, "The first non-padding token of input_ids should be BOS token."
+        )
+        bos_indices = [i for i, token in enumerate(input_ids) if token == self.bos_token_id]
+        self.assertEqual(len(bos_indices), 1, "There should be exactly one BOS token in input_ids.")
```
Comment (qgallouedec) [inline tests/test_utils.py:222]:
```suggestion
        self.assertEqual(input_ids.count(self.bos_token_id), 1, "There should be exactly one BOS token in input_ids.")
```

**Review by qgallouedec — APPROVED** [review]
LGTM, I've just added a minor suggestion

**Comment by ruijunfeng** [comment]
@kashif Sorry I used a wrong version of the code. I have tried it again and the refactored code is all good.

**Pull request merged**